README.md
# RBarchart v0.2.1

[![Build Status](https://travis-ci.org/mariope/RBarchart.png?branch=master)](https://travis-ci.org/mariope/RBarchart)

Barchart OnDemand features a diverse set of market data and information APIs. 
Pricing is based on the number of monthly queries and fields requested. 
They offer four packages: Small, Medium, Large, Enterprise.

## Getting Started

[Barchart.com](https://www.barchart.com/) have available a free API to retrieve quotes and 
historical data and they have a diverse premium set of market data and information APIs. See [Barchart OnDemand](https://www.barchart.com/ondemand)

**RBarchart** package v0.2.1 implements **getQuote**, **getHistory**, **getSymbolLookUp**, **getQuoteEod**, **getClosePrice**, **getFuturesExpirations**, **getFuturesSpecifications** methods from Barchart API.

### Prerequisities

If you install from devtools you do not need to do anything else.

```
devtools::install_github("mariope/RBarchart")
library(RBarchart)
setAPIkey('<Your API key>', premium = FALSE)
```

#### Getting started

When you load the library, you obtain three functions: 

- **setAPIkey** to set the Barchart API key and define if it is premium API.

- **GetHistory** The getHistory API is used to request historical time series data on stocks, indexes, mutual funds, ETFs, futures, foreign exchange, or cryptocurrencies. Historical data is available as tick, minute bars or end-of-day data.

- **getSymbolLookUp** The getSymbolLookUp API provides a list of all instruments based on the keyword being passed.

- **getQuote** The getQuote API is used to request price data, either real-time, delayed or end-of-day, by symbol on stocks, indexes, mutual funds, ETFs, futures, foreign exchange, or cryptocurrencies.

- **getQuoteEod** The getQuoteEod API is used to request end-of-day price data, by combined exchange and symbol, on stocks, indexes, mutual funds, ETFs, futures, foreign exchange, or cryptocurrencies. Historical data is available as tick, minute bars or end-of-day data.

- **getClosePrice** The getClosePrice API provides the close price for given instruments for the given date.

- **getFuturesExpirations** The getFuturesExpiration API from Barchart OnDemand provides first notice and last trade dates for futures contracts.

- **getFuturesSpecifications** Barchart OnDemand's getFuturesSpecifications API provides contract information such as trading hours, contract size and tick size for futures contracts.

```
library(RBarchart)
```

## Running RBarchart

The functions available are:

### setAPIkey

Set the API key and define if it is premium.

```
> setAPIkey('<Your API key>', premium = FALSE)
```

In order to get an API key, please, visit [Barchart Ondemand](https://www.barchart.com/ondemand/api)

### getHistory

The getHistory API is used to request historical time series data on stocks,
indexes, mutual funds, ETFs, futures, foreign exchange, or cryptocurrencies.
Historical data is available as tick, minute bars or end-of-day data.

```
> getHistory(symbol = 'ESM19', type = 'dailyContinue')
```

### getSymbolLookUp

The getSymbolLookUp API provides a list of all instruments based on the keyword being passed.

```
> getSymbolLookUp('Gold', limit = 100, exchanges = 'COMEX')
```

### getQuote

The getQuote API is used to request price data, either real-time, 
delayed or end-of-day, by symbol on stocks, indexes, mutual funds, ETFs,
futures, foreign exchange, or cryptocurrencies.

```
> getQuote(symbols = 'ESM19')
> getQuote(symbols = 'NGM19,ESM19', fields = 'openInterest')
```

### getQuoteEod

The getQuoteEod API is used to request end-of-day price data, by combined exchange and symbol, on stocks, indexes, mutual funds, ETFs, futures, foreign exchange, or cryptocurrencies. Historical data is available as tick, minute bars or end-of-day data.

```
```

### getClosePrice

The getClosePrice API provides the close price for given instruments for the given date.

```
```

### getFuturesExpirations

The getFuturesExpiration API from Barchart OnDemand provides first notice and last trade dates for futures contracts.

```
```

### getFuturesSpecifications

Barchart OnDemand's getFuturesSpecifications API provides contract information such as trading hours, contract size and tick size for futures contracts.

```
```

## Built With

* R v3.3.3
* R Studio - version 1.1.463
* XML package v3.98
* httr package v1.3.1

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

I use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/mariope/RBarchart/tags). 

## Authors

* **Mario Pisa** - *Initial work* - [RBarchart](https://github.com/mariope/RBarchart)

## License

This project is licensed under the GPL-3 License - see the [LICENSE](LICENSE) file for details

## Acknowledgments

* To Open Source community.


NEWS.md
RBarchart v0.2.1 (Release date: 2019-06-22)
================

Changes:

* Added the getSymbolLookUp XML method.
* Added an option to create an eXtended Time Series (xts) output to getHistory function.


RBarchart v0.2.0
================

Changes:

* Added the getFuturesExirations XML method.
* Added the getFuturesSpecifications XML method.
* Added the getClosePrice XML method.
* Added the getQuoteEod XML method.
* Fix the Title and Description in DESCRIPTION file based on CRAN feedback.


RBarchart v0.1.0 (Release date: 2019-04-22)
================

Changes:

* Added the setAPIkey function.
* Added the getQuote XML method.
* Added the getHistory XML method.


R/getQuote.R
#' getQuote Barchart data.
#'
#' \code{getQuote} returns an data frame
#' The getQuote API is used to request price data, either real-time,
#' delayed or end-of-day, by symbol on stocks, indexes, mutual funds, ETFs,
#' futures, foreign exchange, or cryptocurrencies.
#'
#' @author Mario Pisa
#' @param symbols Required; A symbol or code that identifies a financial instrument. Multiple symbols separated by a comma may be used. For futures, notation such as for the active contract or for the first nearby is supported, as is for all futures contracts for a given root symbol, as is for all options contracts for a given underlying futures symbol.
#' @param fields optional; The fields requested. Example fiftyTwoWkHigh,fiftyTwoWkHighDate,fiftyTwoWkLow,fiftyTwoWkLowDate
#' @param only optional; Returns only specified fields. Example symbol,name
#' @return data frame
#' @examples
#' \donttest{
#' getQuote(symbols = 'ESM19')
#' getQuote(symbols = 'NGM19,ESM19', fields = 'openInterest')
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api/getQuote}
#' @export
getQuote <- function(symbols, fields = NULL, only = NULL) {

   # if (!exists("apikey") | !exists("url")) {
   #    print("Please, exec setAPIkey(\"<Your API key>\", premium = FALSE) function")
   #    stop()
   # }

   query <- 'getQuote.xml'
   getMethod <- paste0(getOption("url"),
                       query,
                       '?apikey=', getOption("apikey"),
                       '&symbols=', symbols,
                       '&fields=', fields,
                       '&only=', only)

   res <- httr::GET(getMethod)
   res_df <- XML::xmlToDataFrame(rawToChar(res$content))
   print(paste(res_df[[1, 'code']], res_df[[1, 'message']]))
   res_df <- res_df[-1, -c(1, 2)]
   return(res_df)

}


R/setAPIkey.R
#' setAPIkey Barchart.
#'
#' \code{setAPIkey} returns a hide APIkey variable.
#' The setAPIkey is used to setup the Barchart API key.
#'
#' @author Mario Pisa
#' @param apikey required; Barchart API key
#' @param premium optional; default = FALSE
#' @return var
#' @examples
#' \donttest{
#' setAPIkey(apikey = '<Your Barchart API key>')
#' setAPIkey(apikey = '<Your Barchart API key>', premium = TRUE)
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api}
#' @export
setAPIkey <- function(apikey, premium = FALSE) {

   if (!missing(apikey)) {
      options(apikey = apikey)
   }
   invisible(getOption("apikey"))

   if (isTRUE(premium)) {
      options(url = 'https://ondemand.websol.barchart.com/')
   } else {
      options(url = 'https://marketdata.websol.barchart.com/')
   }
   invisible(getOption("url"))
}


R/getQuoteEod.R
#' getQuoteEod Barchart data.
#'
#' \code{getQuoteEod} returns an data frame
#' The getQuoteEod API is used to request end-of-day price data,
#' by combined exchange and symbol, on stocks, indexes, mutual funds, ETFs,
#' futures, foreign exchange, or cryptocurrencies. Historical data is
#' available as tick, minute bars or end-of-day data.
#'
#' @author Mario Pisa
#' @param symbols required; An MIC code or a Country code plus the symbol or code that identifies a financial instrument. Multiple symbols separated by a comma may be used. For futures, notation such as for the active contract or for the first nearby is supported, as is for all futures contracts for a given root symbol. If no country code or MIC is passed then the API will behave the same as the getQuote API. Example XASE:SPY,US:IBM
#' @param exchange optional; The list of valid exchange codes to limit symbol search. Example NYSE,AMEX
#' @return data frame
#' @examples
#' \donttest{
#' getQuoteEod(symbols = 'XASE:SPY')
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api/getQuoteEod}
#' @export
getQuoteEod <- function(symbols, exchange = NULL) {

   # if (!exists("apikey") | !exists("url")) {
   #    print("Please, exec setAPIkey(\"<Your API key>\", premium = FALSE) function")
   #    stop()
   # }

   query <- 'getQuoteEod.xml'
   getMethod <- paste0(getOption("url"),
                       query,
                       '?apikey=', getOption("apikey"),
                       '&symbols=', symbols,
                       '&exchange=', exchange)

   res <- httr::GET(getMethod)
   res_df <- XML::xmlToDataFrame(rawToChar(res$content))
   print(paste(res_df[[1, 'code']], res_df[[1, 'message']]))
   res_df <- res_df[-1, -c(1, 2)]
   return(res_df)

}


R/getClosePrice.R
#' getClosePrice Barchart data.
#'
#' \code{getClosePrice} returns an data frame
#' The getClosePrice API provides the close price for given instruments for the given date.
#'
#' @author Mario Pisa
#' @param symbols required; A symbol or code that identifies a financial instrument. Multiple symbols separated by a comma may be used. Example IBM,AAPL,GOOG
#' @param date optional; The date of the historical data query. The value should conform to the format yyyymmdd. Example 2010-01-01
#' @param splits optional; An adjustment of stock value due to corporate action. This parameter only applies to stocks and specifies whether the data returned should be adjusted for splits or not. Set to true to query for adjusted the data, or to false for non-adjusted data. Default 1. Example true
#' @param dividends optional; A distribution of a portion of a company's earnings. This parameter only applies to stocks and specifies whether the data returned should be adjusted for dividends or not. Set to true to query for adjusted the data, or to false for non-adjusted data. Default 1. Example true
#' @return data frame
#' @examples
#' \donttest{
#' getClosePrice(symbols = 'ESM19')
#' getClosePrice(symbols = 'NGM19,ESM19')
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api/getFuturesExpirations}
#' @export
getClosePrice <- function(symbols, date = NULL, splits = 1, dividends = 1) {

   # if (!exists("apikey") | !exists("url")) {
   #    print("Please, exec setAPIkey(\"<Your API key>\", premium = FALSE) function")
   #    stop()
   # }

   query <- 'getClosePrice.xml'
   getMethod <- paste0(getOption("url"),
                       query,
                       '?apikey=', getOption("apikey"),
                       '&symbols=', symbols,
                       '&date=', date,
                       '&splits=', splits,
                       '&dividends=', dividends)

   res <- httr::GET(getMethod)
   res_df <- XML::xmlToDataFrame(rawToChar(res$content))
   print(paste(res_df[[1, 'code']], res_df[[1, 'message']]))
   res_df <- res_df[-1, -c(1, 2)]
   return(res_df)

}


R/getSymbolLookup.R
#' getSymbolLookup Barchart data.
#'
#' \code{getSymbolLookUp} returns an data frame.
#' The getSymbolLookUp API provides a list of all instruments based on the keyword being passed.
#'
#' @author Mario Pisa
#' @param keyword Required; The keyword on which the search needs to be made.
#' @param limit optional; The number of results to be returned.
#' @param exchanges optional; Valid exchange code. Multiple exchanges can be passed separated by a comma. Default NYSE,AMEX,NASDAQ,COMEX
#' @param moreBestMatches optional; To get additional best match results. Boolean. Dafault = 'true'.
#' @param moreBestMatchLimit optional; The number of additional best match results that should be returned. Default = 4
#' @param fields optional; The fields requested. A comma or semi-colon delimited string.
#'
#' @return data frame
#' @examples
#' \donttest{
#' getSymbolLookUp('GC', limit = 100, exchanges = 'COMEX')
#' getSymbolLookUp('Gold', limit = 100, exchanges = 'COMEX')
#' getSymbolLookUp('Gold', limit = 100, exchanges = 'CXMI')
#' getSymbolLookUp('Zinc', limit = 1000, exchanges = 'COMEX')
#' getSymbolLookUp('A', limit = 1000, exchanges = 'COMEX')
#' getSymbolLookUp('E-Mini', limit = 100, exchanges = 'CXMI')
#' getSymbolLookUp('E-Micro', limit = 100, exchanges = 'CXMI')
#' getSymbolLookUp('19', limit = 100, exchanges = 'CXMI')
#' getSymbolLookUp('19', limit = 100, exchanges = 'COMEX')
#' getSymbolLookUp('%', limit = 90000, exchanges = 'COMEX', moreBestMatches = 'true')
#' getSymbolLookUp('%', limit = 90000, exchanges = 'CXMI')
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api/getSymbolLookUp}
#' @export
getSymbolLookUp <- function(keyword, limit = 10, exchanges = 'NYSE,AMEX,NASDAQ,COMEX',
                            moreBestMatches = 'true', moreBestMatchLimit = 4, fields = '') {

   query <- 'getSymbolLookUp.xml'
   getMethod <- paste0(getOption("url"),
                       query,
                       '?apikey=', getOption("apikey"),
                       '&keyword=', keyword,
                       '&limit=', limit,
                       '&exchanges=', exchanges,
                       '&moreBestMatches=', moreBestMatches,
                       '&moreBestMatchLimit=', moreBestMatchLimit,
                       '&fields=', fields
   )

   res <- httr::GET(getMethod)
   res_df <- XML::xmlToDataFrame(rawToChar(res$content))
   print(paste(res_df[[1, 'code']], res_df[[1, 'message']]))
   res_df <- res_df[-1, -c(1, 2)]
   rownames(res_df) <- NULL
   return(res_df)

}


R/getFuturesExpirations.R
#' getFuturesExpirations Barchart data.
#'
#' \code{getFuturesExpirations} returns an data frame
#' The getFuturesExpiration API from Barchart OnDemand provides first notice and last trade dates for futures contracts.
#'
#' @author Mario Pisa
#' @param roots optional; A symbol or code that identifies a futures root. Example ZC,ZS
#' @param contracts optional; A symbol or code that identifies a futures contract.
#' @return data frame
#' @examples
#' \donttest{
#' getFuturesExpirations(roots = 'ES')
#' getFuturesExpirations(contracts = 'NGM19,ESM19')
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api/getFuturesExpirations}
#' @export
getFuturesExpirations <- function(roots = NULL, contracts = NULL) {

   # if (!exists("apikey") | !exists("url")) {
   #    print("Please, exec setAPIkey(\"<Your API key>\", premium = FALSE) function")
   #    stop()
   # }

   query <- 'getFuturesExpirations.xml'
   getMethod <- paste0(getOption("url"),
                       query,
                       '?apikey=', getOption("apikey"),
                       '&roots=', roots,
                       '&contracts=', contracts)

   res <- httr::GET(getMethod)
   res_df <- XML::xmlToDataFrame(rawToChar(res$content))
   print(paste(res_df[[1, 'code']], res_df[[1, 'message']]))
   res_df <- res_df[-1, -c(1, 2)]
   return(res_df)

}


R/getFuturesSpecifications.R
#' getFuturesSpecifications Barchart data.
#'
#' \code{getFuturesSpecifications} returns an data frame
#' Barchart OnDemand's getFuturesSpecifications API provides contract information such as trading hours, contract size and tick size for futures contracts.
#'
#' @author Mario Pisa
#' @param symbols optional; A symbol or code that identifies a financial instrument. Multiple symbols separated by a comma may be used. Example GC,CL
#' @param exchanges optional; The code for the exchange a group of financial instruments are listed on. Example COMEX,NYMEX
#' @param futuresCategory optional; Futures categories include: Energies, Financials, Grains, Indices, Meats, Metals, Softs and Currencies. Example energies,metals
#' @return data frame
#' @examples
#' \donttest{
#' getFuturesSpecifications(symbols = 'CL')
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api/getFuturesSpecifications}
#' @export
getFuturesSpecifications <- function(symbols, exchanges = NULL, futuresCategory = NULL) {

   # if (!exists("apikey") | !exists("url")) {
   #    print("Please, exec setAPIkey(\"<Your API key>\", premium = FALSE) function")
   #    stop()
   # }

   query <- 'getFuturesSpecifications.xml'
   getMethod <- paste0(getOption("url"),
                       query,
                       '?apikey=', getOption("apikey"),
                       '&symbols=', symbols,
                       '&exchanges=', exchanges,
                       '&futuresCategory=', futuresCategory)

   res <- httr::GET(getMethod)
   res_df <- XML::xmlToDataFrame(rawToChar(res$content))
   print(paste(res_df[[1, 'code']], res_df[[1, 'message']]))
   res_df <- res_df[-1, -c(1, 2)]
   return(res_df)

}


R/getHistory.R
#' getHistory Barchart data.
#'
#' \code{getHistory} returns an data frame.
#' The getHistory API is used to request historical time series data on stocks,
#' indexes, mutual funds, ETFs, futures, foreign exchange, or cryptocurrencies.
#' Historical data is available as tick, minute bars or end-of-day data.
#'
#' @author Mario Pisa
#' @param symbol Required; A symbol or code that identifies a financial instrument.
#' @param type Required; The type of historical data to return, including tick data, minute data, and end-of-day data. Valid Values: ticks, minutes, nearbyMinutes, formTMinutes, daily, dailyNearest, dailyContinue, weekly, weeklyNearest, weeklyContinue, monthly, monthlyNearest, monthlyContinue, quarterly, quarterlyNearest, quarterlyContinue, yearly, yearlyNearest, yearlyContinue. Example minutes
#' @param startDate optional; The start date of the historical data query. This parameter should be set to the desired start date/time for the query (the result set will include records back to, and including, this value). If not set, the value will default to the beginning of the day specified in the end parameter, if end is specified, or to the beginning of the current day, if end is not specified. The value should conform to the format yyyymmdd[hhmm[ss]], where fields in brackets are optional (Do not include the brackets themselves). Any optional fields that are not explicitly set will default to 0 (i.e. 20090203 will default to 20090203000000 or February 3, 2009 at 00:00:00).
#' @param endDate optional; The end data of the historical data query. This parameter should be set to the desired end date/time for the query (the result set will include records up to, but not including, this value). If not set, the value will default to the end of the day specified in the start parameter, if specified, or to the end of the current day, if start is not specified. The value should conform to the format yyyymmdd[hhmm[ss]], where fields in brackets are optional (Do not include the brackets themselves). Any optional fields that are not explicitly set will default to 0 (i.e. 20090203 will default to 20090203000000 or February 3, 2009 at 00:00:00).
#' @param maxRecords optional; The maximum amount of records returned. This parameter should be set to the maximum number of records desired. If not specified, there number of records returned will be determined by the date/time parameters specified as well as any defaults that apply to the query.
#' @param interval optional; The number of minutes for a minute query.
#' @param order optional; An arrangement of fields within a particular record (ascending or descending). This parameter can be set to one of two values ("asc" and "desc") in order to specify the chronological order of the result set returned. If this parameter is not specified, the order results is not guaranteed.
#' @param sessionFilter optional; This parameter modifies the default session codes/sale conditions used to return ticks for each exchange. For NYSE and AMEX, the default session filter is "@EFKX56V9" (meaning all ticks with sale conditions corresponding to one of the characters in the filter are included in the results), for NASDAQ the default is "@ABDEFKOSXY156", and for everything else all session codes/sale conditions are returned except the settle (session code '*'). If the session filter is set to a string of valid session codes (i.e. "EFK"), only ticks with the specified session codes are included in the results. If the string is prefixed with character '!' (i.e."!EFK"), all session codes except those in the string are included in the results. If the string is prefixed with character '+' (i.e. "+T"), then all the default session codes in addition to the ones specified in the string are included in the results. And if the string is prefixed with character '-' (i.e. "-EF") then all default session codes except the ones specified are included in the results. Please note that the '+' character should be escaped (to \%2B) when entering the URL in a web browser or executing the query in an API that does not escape it by default.
#' @param splits optional; An adjustment of stock value due to corporate action. This parameter only applies to stocks and specifies whether the data returned should be adjusted for splits or not. Set to true to query for adjusted the data, or to false for non-adjusted data. If not specified, the default is true. In order to guarantee the same adjustment settings in the future, this parameter should be specified.
#' @param dividends optional; A distribution of a portion of a company's earnings. This parameter only applies to stocks and specifies whether the data returned should be adjusted for dividends or not. Set to true to query for adjusted the data, or to false for non-adjusted data. If not specified, the default is true. In order to guarantee the same adjustment settings in the future, this parameter should be specified.
#' @param volume optional; The quantity of shares or contracts traded. For futures, this parameter can be set to one of two values (contract and total) in order to specify whether the volume returned should be the contract volume or the total volume. For aggregates (such as weekly, monthly or yearly), this returns the average volume for the period specified. If the value is preceded by 'sum' (sumcontract and sumtotal), then it returns the sum of the volumes in each daily bar during the period specified. If not specified, the value will default to contract. For aggregate equities queries (such as weekly, monthly or yearly), this parameter can be set to sum to return the sum of the volumes in each daily bar during the period specified. If not specified, then the average volume is returned. Valid Values: total, sum, contract, sumcontract, sumtotal
#' @param nearby optional; This parameter specifies the offset from the front month for 'nearest' queries (data parameter set to dailynearest, weeklynearest, monthlynearest, quarterlynearest and yearlynearest). The default value for this parameter is 1, which sets nearest queries to the most current front month. If set to a value greater than 1, then the nth front month is used (for example, in August of 2010, ESU10 would be the current front month, so nearby=2 would use ESZ10, nearby=3 would use ESH11, etc.) This parameter is ignored for all other queries. Alternatively, the same functionality provided by the nearby parameter can be achieved using the symbol notation symbol=RS*n, where RS is the root symbol and n is the nth front month (i.e. symbol=ES*1, symbol=YM*3, etc.). When using this notation on the symbol parameter, the nearby parameter should be omitted. Example 1
#' @param jerq optional; Whether to merge with JERQ data. Default 1. Example true
#' @param exchange optional; The list of valid exchange codes to limit symbol search. Default NYSE,AMEX,NASDAQ,COMEX
#' @param backAdjust optional; This parameter specifies whether the contracts in the series will be adjusted based on the roll-gap between the closing prices of the current contract and the previous contract on the day of the switch. Valid values are true and false. If not specified, the default value is false resulting in a non-adjusted query. This parameter only applies to multi-contract futures queries and is ignored for all other queries. Valid Values: true, false. Default false
#' @param daysToExpiration optional; This parameter specifies the number of calendar days prior to a contract expiration when the series of contracts will be switched to the next contract in the series. Valid values are 0 through 60. A value of 0 will tell the system to use the day of expiration. If not specified, the default value (1) is used, resulting in each contract in the series to run until (and including) the day prior to its expiration. This parameter only applies to multi-contract futures queries and is ignored for all other queries. Valid Values: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60. Default 1. Example 1.
#' @param contractRoll optional; For futures, multi-contract nearest queries (data parameter set to dailynearest, weeklynearest, monthlynearest, quarterlynearest or yearlynearest), this parameter can be set to one of two values (expiration, combined), and determines how the switch from one contract to the next in the series is calculated. When the value specified is 'expiration', the switch from one contract to the next in the series will be based on the expiration date (and the value of the 'daysToExpiration' parameter if specified). When the value specified is 'combined', a combination of volume and openinterest will be used to determine when to switch from one contract to the next in the series (when using this value, the 'daysToExpiration' parameter is ignored). Valid Values: expiration, combined. Default expiration. Example expiration
#' @param output_type optional; options('df' | 'xts') default= 'df'
#' @return data frame
#' @examples
#' \donttest{
#' getHistory(symbol = 'ESM19', type = 'dailyContinue')
#' getHistory(symbol = 'ESM19', type = 'dailyContinue', daysToExpiration = 8)
#' getHistory(symbol = 'ESM19', type = 'dailyContinue', output_type = 'xts')
#' getHistory(symbol = 'ESM19', type = 'minutes')
#' getHistory(symbol = 'ESM19', type = 'formTMinutes', interval = 30)
#' getHistory(symbol = 'ESM19', type = 'dailyContinue', startDate = '20100101')
#' getHistory(symbol = 'ESM19', type = 'daily', startDate = '20190101', endDate = '20190422')
#' getHistory(symbol = 'GCQ19', type = 'dailyContinue', startDate = '20100101',
#'    endDate = '20190621', exchange = 'COMEX', daysToExpiration = 8)
#' getHistory(symbol = 'GCQ19', type = 'ticks', startDate = '20190621 10:00:00',
#'    endDate = '20190621 10:05:00', exchange = 'COMEX')
#' }
#' @seealso \url{https://www.barchart.com/ondemand/api/getHistory}
#' @export
getHistory <- function(symbol, type, startDate = NULL, endDate = NULL, maxRecords = '1000000',
                       interval = 1, order = 'asc', sessionFilter = NULL, splits = 1,
                       dividends = 1, volume = 'contract', nearby = 1, jerq = 1,
                       exchange = 'NYSE,AMEX,NASDAQ,COMEX', backAdjust = 'false',
                       daysToExpiration = 1, contractRoll = 'expiration',
                       output_type = 'df') {

   # if (!exists("apikey") | !exists("url")) {
   #    warning("Please, exec setAPIkey(\"<Your API key>\", premium = FALSE) function")
   #    stop()
   # }

   query <- 'getHistory.xml'
   getMethod <- paste0(getOption("url"),
                       query,
                       '?apikey=', getOption("apikey"),
                       '&symbol=', symbol,
                       '&type=', type,
                       '&startDate=', startDate,
                       '&endDate=', endDate,
                       '&maxRecords=', maxRecords,
                       '&interval=', interval,
                       '&order=', order,
                       '&sessionFilter=', sessionFilter,
                       '&splits=', splits,
                       '&dividends=', dividends,
                       '&volume=', volume,
                       '&nearby=', nearby,
                       '&jerq=', jerq,
                       '&exchange=', exchange, # 'NYSE%2CAMEX%2CNASDAQ',
                       '&backAdjust=', backAdjust,
                       '&daysToExpiration=', daysToExpiration,
                       '&contractRoll=', contractRoll
                       )

   res <- httr::GET(getMethod)
   res_df <- XML::xmlToDataFrame(rawToChar(res$content))
   print(paste(res_df[[1, 'code']], res_df[[1, 'message']]))
   res_df <- res_df[-1, -c(1, 2)]
   rownames(res_df) <- NULL

   if (nrow(res_df) > 0 && output_type == 'xts') {
      res_df <- xts::as.xts(res_df[, c(-(1:3))],
                            order.by = as.POSIXct(res_df[, 2], format = "%Y-%m-%dT%H:%M:%S"))
      storage.mode(res_df) <- 'numeric'
      res_df <- zoo::na.locf(res_df)
   }

   return(res_df)

}
